kubernetes/oauth2-proxy.jsonnet

```diff
@@ -0,0 +1,115 @@
+    secret=error 'secret must be set',
+
+    // domain is the fully qualified domain name that the oauth2-proxy
+    // should use for the cookie domain. This is required.
+    domain=error 'domain must be set',
+
+    // listenPort is the port that the oauth2-proxy should listen on.
+    listenPort=8080,
+
+    // oidceIssuerURL is the URL of the OIDC issuer.
+    oidcIssuerURL='https://outreach.okta.com'
+  ):: ok.Container('oauth2-proxy') {
+    // Map to listen_port so we can easily access this in other functions
+    listen_port:: listenPort,
+
+    image: 'gcr.io/outreach-docker/quay.io/oauth2-proxy/oauth2-proxy:v7.3.0',
+    args: [
+      '--upstream=http://localhost:%d/' % servicePort,
+      '--provider=oidc',
+      '--cookie-name=%s_oauth2_proxy' % service.metadata.name,
+      '--cookie-secure=true',
+      '--cookie-expire=168h0m',
+      '--cookie-refresh=8h',
+      '--cookie-domain=' + domain,
+      '--http-address=0.0.0.0:%d' % $.listen_port,
+      '--oidc-issuer-url=%s' % oidcIssuerURL,
+      '--redirect-url=https://%s/oauth2/callback' % domain,
+      '--email-domain=outreach.io',
+      '--pass-access-token=true',
+      '--skip-provider-button=true',
+      '--request-logging=false',
+      '--silence-ping-logging=true',
+    ],
+    envFrom: [
+      { secretRef: { name: secret.metadata.name } },
+    ],
+    ports_:: {
+      'oauth2-proxy': 8080,
+    },
+    resources: {
+      limits: self.requests {
+        memory: '200Mi',
+      },
+      requests: {
+        cpu: '10m',
+        memory: '110Mi',
+      },
+    },
+  },
+
+  // ServicePort returns an entry for the oauth2-proxy container to be
+  // used in a Service. The only required argument is 'container' which
+  // is the oauth2-proxy container created by Container().
+  ServicePort(
+    // container is the oauth2-proxy container created by Container().
+    container=error 'container must be set',
+
+    // port is the port that the oauth2-proxy should use for the service
+    // (this is not the container port)
+    port=80,
+  ):: {
+    port: port,
+    targetPort: container.listen_port,
+    protocol: 'TCP',
+    name: 'http-oauth2-proxy',
+  },
+
+  // IngressRule returns an entry for the oauth2-proxy container to be
+  // used in an Ingress.
+  IngressRule(
+    // container is the oauth2-proxy container created by Container().
+    container=error 'container must be set',
+
+    // serviceName is the name of the Kubernetes service that the oauth2-proxy
+    // is on.
+    serviceName=error 'serviceName must be set',
+  ):: {
+    path: '/*',
+    backend: {
+      serviceName: serviceName,
+      servicePort: 'http',
+    },
+  },
+
+  // ALBIngressAnnotations returns a set of annotations that are helpful when
+  // using the oauth2-proxy with an ALB Ingress Controller.
+  ALBIngressAnnotations():: {
+    'alb.ingress.kubernetes.io/load-balancer-attributes'+: ',routing.http.preserve_host_header.enabled=true',
```

Comment:
I'm curious about what this usage will look like.

Reply:
The usage would be `existingAnnotations + ALBIngressAnnotations`, right?

Reply:
That looks like it works in the jsonnet repl.